frontend/include/chpl/resolution/resolution-types.h

```diff
@@ -1066,8 +1066,141 @@ class ApplicabilityResult {
+  int actualIdx() const { return actualIdx_; }
+  bool hasActual() const { return hasActual_; }
+  bool formalInstantiated() const { return formalInstantiated_; }
+  bool hasDefault() const { return hasDefault_; }
+  bool isVarArgEntry() const { return isVarArgEntry_; }
+};
+
+/** FormalActualMap maps formals to actuals */
+class FormalActualMap {
+ private:
+  std::vector<FormalActual> byFormalIdx_;
+  std::vector<int> actualIdxToFormalIdx_;
+  bool mappingIsValid_ = false;
+  int failingActualIdx_ = -1;
+  int failingFormalIdx_ = -1;
+
+ public:
+
+  using FormalActualIterable = Iterable<std::vector<FormalActual>>;
+
+  FormalActualMap(const UntypedFnSignature* sig, const CallInfo& call) {
+    mappingIsValid_ = computeAlignment(sig, nullptr, call);
+  }
+  FormalActualMap(const TypedFnSignature* sig, const CallInfo& call) {
+    mappingIsValid_ = computeAlignment(sig->untyped(), sig, call);
+  }
+
+  bool operator==(const FormalActualMap& other) const {
+    return byFormalIdx_ == other.byFormalIdx_ &&
+           actualIdxToFormalIdx_ == other.actualIdxToFormalIdx_ &&
+           mappingIsValid_ == other.mappingIsValid_ &&
+           failingActualIdx_ == other.failingActualIdx_ &&
+           failingFormalIdx_ == other.failingFormalIdx_;
+  }
+
+  bool operator!=(const FormalActualMap& other) const {
+    return !(*this == other);
+  }
+
+  void mark(Context* context) const {
+    for (auto const &elt : byFormalIdx_) {
+      elt.mark(context);
+    }
+    (void) actualIdxToFormalIdx_; // nothing to mark
+    (void) mappingIsValid_; // nothing to mark
+    (void) failingActualIdx_; // nothing to mark
+    (void) failingFormalIdx_; // nothing to mark
+  }
+
+  size_t hash() const {
+    return chpl::hash(byFormalIdx_, actualIdxToFormalIdx_, mappingIsValid_,
+                      failingActualIdx_, failingFormalIdx_);
+  }
+
+  /** check if mapping is valid */
+  bool isValid() const { return mappingIsValid_; }
+
+  /** get the FormalActuals in the order of the formal arguments */
+  FormalActualIterable byFormals() const {
+    return FormalActualIterable(byFormalIdx_);
+  }
+
+  /** get the FormalActual for a particular formal index */
+  const FormalActual& byFormalIdx(int formalIdx) const {
+    CHPL_ASSERT(0 <= formalIdx && (size_t) formalIdx < byFormalIdx_.size());
+    return byFormalIdx_[formalIdx];
+  }
+
+  /** get the FormalActual for a particular actual index,
+      and returns nullptr if none was found. */
+  const FormalActual* byActualIdx(int actualIdx) const {
+    if (actualIdx < 0 || (size_t) actualIdx >= actualIdxToFormalIdx_.size())
+      return nullptr;
+    int formalIdx = actualIdxToFormalIdx_[actualIdx];
+    if (formalIdx < 0 || (size_t) formalIdx >= byFormalIdx_.size())
+      return nullptr;
+    return &byFormalIdx_[formalIdx];
+  }
+
+ private:
+  bool computeAlignment(const UntypedFnSignature* untyped,
+                        const TypedFnSignature* typed, const CallInfo& call);
+};
```

Comment:
Did you change anything in FormalActualMap / FormalActual other than moving the code?

Generally speaking, I'd like to move away from having `resolution-types.h` store miscellaneous types. So, I'd rather you move these to their own header, for example, frontend/include/chpl/resolution/FormalActualMap.h`.

Reply:
I think this might be better suited for a separate PR. I tried moving `FormalActualMap`, but that would require moving `UntypedFnSignature`, and maybe others...

Reply:
And yes, I did change FormalActualMap a bit -- added some missing methods for the query framework. Perhaps == and hash.